cmd/util/ledger/migrations/staged_contracts_migration.go

```diff
@@ -0,0 +1,207 @@
+		m.contracts[address] = map[flow.RegisterID]Contract{}
+	}
+
+	registerID := flow.ContractRegisterID(flow.ConvertAddress(address), change.name)
+
+	_, exist := m.contracts[address][registerID]
+	if exist {
+		// Staged multiple updates for the same contract.
+		// Overwrite the previous update.
+		m.log.Warn().Msgf(
+			"existing staged update found for contract %s.%s. Previous update will be overwritten.",
+			address.HexWithPrefix(),
+			change.name,
+		)
+	}
+
+	m.contracts[address][registerID] = change.Contract
+}
+
+func (m *StagedContractsMigration) contractUpdatesForAccount(
+	address common.Address,
+) (map[flow.RegisterID]Contract, bool) {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	contracts, ok := m.contracts[address]
+
+	// remove address from set of addresses
+	// to keep track of which addresses are left to change
+	delete(m.contracts, address)
+
+	return contracts, ok
+}
+
+func (m *StagedContractsMigration) MigrateAccount(
+	_ context.Context,
+	address common.Address,
+	payloads []*ledger.Payload,
+) ([]*ledger.Payload, error) {
+
+	contractUpdates, ok := m.contractUpdatesForAccount(address)
+	if !ok {
+		// no contracts to change on this address
+		return payloads, nil
+	}
+
+	for payloadIndex, payload := range payloads {
+		key, err := payload.Key()
+		if err != nil {
+			return nil, err
+		}
+
+		registerID, err := convert.LedgerKeyToRegisterID(key)
+		if err != nil {
+			return nil, err
+		}
+
+		updatedContract, ok := contractUpdates[registerID]
+		if !ok {
+			// not a contract register, or
+			// not interested in this contract
+			continue
+		}
+
+		name := updatedContract.name
+		newCode := updatedContract.code
+		oldCode := payload.Value()
+
+		err = m.checkUpdateValidity(name, newCode, oldCode)
+		if err != nil {
+			m.log.Error().Err(err).
+				Msgf(
+					"fail to update contract %s in account %s",
+					name,
+					address.HexWithPrefix(),
+				)
+		} else {
+			// change contract code
+			payloads[payloadIndex] = ledger.NewPayload(
+				key,
+				[]byte(newCode),
+			)
+		}
+
+		// remove contract from list of contracts to change
+		// to keep track of which contracts are left to change
+		delete(contractUpdates, registerID)
+	}
+
+	if len(contractUpdates) > 0 {
+		return nil, fmt.Errorf("failed to find all contract registers that need to be changed")
+	}
+
+	return payloads, nil
+}
+
+func (m *StagedContractsMigration) checkUpdateValidity(contractName, newCode string, oldCode ledger.Value) error {
+	newProgram, err := parser.ParseProgram(nil, []byte(newCode), parser.Config{})
+	if err != nil {
+		return err
+	}
+
+	// TODO: Use the old parser
+	oldProgram, err := parser.ParseProgram(nil, oldCode, parser.Config{})
+	if err != nil {
+		return err
+	}
+
+	// TODO: Use new update validator
+	validator := stdlib.NewContractUpdateValidator(
+		nil,
+		contractName,
+		oldProgram,
+		newProgram,
+	)
+
+	return validator.Validate()
+}
+
+func GetStagedContracts() []StagedContracts {
+	// TODO:
+	return nil
```

Comment:
TBD: To be constructed from the list/csv obtained from the staged contract deployment. See: https://discord.com/channels/613813861610684416/1108479699732152503/1200140480269254696

Reply:
Can you please create an issue for that remainig work? We can definitely get this PR in before, as it depends on Gio's work on the staged contract update contract. Maybe define an "interface"/format – maybe we can read in JSON-CDC?

Reply:
https://github.com/onflow/cadence/issues/3050